nix 0.24 made VsockAddr opaque

https://github.com/nix-rust/nix/pull/1614

vsock won't build with nix >= 0.24.

<details>
<summary>Build log</summary>

```
error[E0423]: cannot initialize a tuple struct which contains private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:117:32
     |
117  |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
     |                                ^^^^^^^^^ help: a local variable with a similar name exists: `vsock_addr`
     |
note: constructor is not visible here due to private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/nix-0.24.1/src/sys/socket/addr.rs:2518:26
     |
2518 |     pub struct VsockAddr(pub(in super::super) sockaddr_vm);
     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ private field

error[E0423]: cannot initialize a tuple struct which contains private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:281:32
     |
281  |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
     |                                ^^^^^^^^^ help: a local variable with a similar name exists: `vsock_addr`
     |
note: constructor is not visible here due to private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/nix-0.24.1/src/sys/socket/addr.rs:2518:26
     |
2518 |     pub struct VsockAddr(pub(in super::super) sockaddr_vm);
     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ private field

error[E0423]: cannot initialize a tuple struct which contains private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:305:32
     |
305  |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
     |                                ^^^^^^^^^ help: a local variable with a similar name exists: `vsock_addr`
     |
note: constructor is not visible here due to private fields
    --> /tmp/tmp.xk1r9Xh1Kg/registry/nix-0.24.1/src/sys/socket/addr.rs:2518:26
     |
2518 |     pub struct VsockAddr(pub(in super::super) sockaddr_vm);
     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ private field

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:29:28
   |
29 | pub use nix::sys::socket::{SockAddr, VsockAddr};
   |                            ^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:57:24
   |
57 |     pub fn bind(addr: &SockAddr) -> Result<VsockListener> {
   |                        ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:58:47
   |
58 |         let mut vsock_addr = if let SockAddr::Vsock(addr) = addr {
   |                                               ^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:94:31
   |
94 |         Self::bind(&SockAddr::Vsock(VsockAddr::new(cid, port)))
   |                               ^^^^^

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:98:40
   |
98 |     pub fn local_addr(&self) -> Result<SockAddr> {
   |                                        ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:117:26
    |
117 |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
    |                          ^^^^^

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:127:50
    |
127 |     pub fn accept(&self) -> Result<(VsockStream, SockAddr)> {
    |                                                  ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:149:27
    |
149 |                 SockAddr::Vsock(VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port)),
    |                           ^^^^^

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:228:27
    |
228 |     pub fn connect(addr: &SockAddr) -> Result<Self> {
    |                           ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:229:43
    |
229 |         let vsock_addr = if let SockAddr::Vsock(addr) = addr {
    |                                           ^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:258:34
    |
258 |         Self::connect(&SockAddr::Vsock(VsockAddr::new(cid, port)))
    |                                  ^^^^^

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:262:39
    |
262 |     pub fn peer_addr(&self) -> Result<SockAddr> {
    |                                       ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:281:26
    |
281 |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
    |                          ^^^^^

warning: use of deprecated enum `nix::sys::socket::SockAddr`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:286:40
    |
286 |     pub fn local_addr(&self) -> Result<SockAddr> {
    |                                        ^^^^^^^^

warning: use of deprecated variant `nix::sys::socket::SockAddr::Vsock`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:305:26
    |
305 |             Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
    |                          ^^^^^

warning: use of deprecated field `nix::sys::socket::SockAddr::Vsock::0`: use SockaddrLike or SockaddrStorage instead
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:58:53
   |
58 |         let mut vsock_addr = if let SockAddr::Vsock(addr) = addr {
   |                                                     ^^^^

error[E0616]: field `0` of struct `VsockAddr` is private
  --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:59:18
   |
59 |             addr.0
   |                  ^ private field

warning: use of deprecated field `nix::sys::socket::SockAddr::Vsock::0`: use SockaddrLike or SockaddrStorage instead
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:229:49
    |
229 |         let vsock_addr = if let SockAddr::Vsock(addr) = addr {
    |                                                 ^^^^

error[E0616]: field `0` of struct `VsockAddr` is private
   --> /tmp/tmp.xk1r9Xh1Kg/registry/vsock-0.2.4/src/lib.rs:230:18
    |
230 |             addr.0
    |                  ^ private field
```

</details>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,7 +12,7 @@ exclude = ["test_fixture"]
 
 [dependencies]
 libc = "0.2.79"
-nix = "0.23.0"
+nix = "0.24.1"
 
 [dev-dependencies]
 rand = "0.8.3"

--- a/echo_server/src/main.rs
+++ b/echo_server/src/main.rs
@@ -19,7 +19,7 @@ use std::io::Read;
 use std::io::Write;
 use std::net::Shutdown;
 use std::thread;
-use vsock::{SockAddr, VsockAddr, VsockListener};
+use vsock::{VsockAddr, VsockListener};
 
 const BLOCK_SIZE: usize = 16384;
 
@@ -48,11 +48,8 @@ fn main() {
         .parse::<u32>()
         .expect("port must be a valid integer");
 
-    let listener = VsockListener::bind(&SockAddr::Vsock(VsockAddr::new(
-        libc::VMADDR_CID_ANY,
-        listen_port,
-    )))
-    .expect("bind and listen failed");
+    let listener = VsockListener::bind(&VsockAddr::new(libc::VMADDR_CID_ANY, listen_port))
+        .expect("bind and listen failed");
 
     println!("Server listening for connections on port {}", listen_port);
     for stream in listener.incoming() {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -18,7 +18,7 @@
 //! Virtio socket support for Rust.
 
 use libc::*;
-use nix::ioctl_read_bad;
+use nix::{ioctl_read_bad, sys::socket::AddressFamily};
 use std::ffi::c_void;
 use std::fs::File;
 use std::io::{Error, ErrorKind, Read, Result, Write};
@@ -28,7 +28,7 @@ use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
 use std::time::Duration;
 
 pub use libc::{VMADDR_CID_ANY, VMADDR_CID_HOST, VMADDR_CID_HYPERVISOR, VMADDR_CID_LOCAL};
-pub use nix::sys::socket::{SockAddr, VsockAddr};
+pub use nix::sys::socket::{SockaddrLike, VsockAddr};
 
 fn new_socket() -> libc::c_int {
     unsafe { socket(AF_VSOCK, SOCK_STREAM | SOCK_CLOEXEC, 0) }
@@ -56,28 +56,20 @@ pub struct VsockListener {
 
 impl VsockListener {
     /// Create a new VsockListener which is bound and listening on the socket address.
-    pub fn bind(addr: &SockAddr) -> Result<VsockListener> {
-        let mut vsock_addr = if let SockAddr::Vsock(addr) = addr {
-            addr.as_ref().to_owned()
-        } else {
+    pub fn bind(addr: &impl SockaddrLike) -> Result<Self> {
+        if addr.family() != Some(AddressFamily::Vsock) {
             return Err(Error::new(
                 ErrorKind::Other,
                 "requires a virtio socket address",
             ));
-        };
+        }
 
         let socket = new_socket();
         if socket < 0 {
             return Err(Error::last_os_error());
         }
 
-        let res = unsafe {
-            bind(
-                socket,
-                &mut vsock_addr as *mut _ as *mut sockaddr,
-                size_of::<sockaddr_vm>() as socklen_t,
-            )
-        };
+        let res = unsafe { bind(socket, addr.as_ptr(), addr.len()) };
         if res < 0 {
             return Err(Error::last_os_error());
         }
@@ -93,11 +85,11 @@ impl VsockListener {
 
     /// Create a new VsockListener with specified cid and port.
     pub fn bind_with_cid_port(cid: u32, port: u32) -> Result<VsockListener> {
-        Self::bind(&SockAddr::Vsock(VsockAddr::new(cid, port)))
+        Self::bind(&VsockAddr::new(cid, port))
     }
 
     /// The local socket address of the listener.
-    pub fn local_addr(&self) -> Result<SockAddr> {
+    pub fn local_addr(&self) -> Result<VsockAddr> {
         let mut vsock_addr = sockaddr_vm {
             svm_family: AF_VSOCK as sa_family_t,
             svm_reserved1: 0,
@@ -116,10 +108,7 @@ impl VsockListener {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr::new(
-                vsock_addr.svm_cid,
-                vsock_addr.svm_port,
-            )))
+            Ok(VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port))
         }
     }
 
@@ -129,7 +118,7 @@ impl VsockListener {
     }
 
     /// Accept a new incoming connection from this listener.
-    pub fn accept(&self) -> Result<(VsockStream, SockAddr)> {
+    pub fn accept(&self) -> Result<(VsockStream, VsockAddr)> {
         let mut vsock_addr = sockaddr_vm {
             svm_family: AF_VSOCK as sa_family_t,
             svm_reserved1: 0,
@@ -151,7 +140,7 @@ impl VsockListener {
         } else {
             Ok((
                 unsafe { VsockStream::from_raw_fd(socket as RawFd) },
-                SockAddr::Vsock(VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port)),
+                VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port),
             ))
         }
     }
@@ -230,41 +219,32 @@ pub struct VsockStream {
 
 impl VsockStream {
     /// Open a connection to a remote host.
-    pub fn connect(addr: &SockAddr) -> Result<Self> {
-        let vsock_addr = if let SockAddr::Vsock(addr) = addr {
-            addr.as_ref()
-        } else {
+    pub fn connect(addr: &impl SockaddrLike) -> Result<Self> {
+        if addr.family() != Some(AddressFamily::Vsock) {
             return Err(Error::new(
                 ErrorKind::Other,
                 "requires a virtio socket address",
             ));
-        };
+        }
 
         let sock = new_socket();
         if sock < 0 {
             return Err(Error::last_os_error());
         }
-        if unsafe {
-            connect(
-                sock,
-                vsock_addr as *const _ as *const sockaddr,
-                size_of::<sockaddr_vm>() as socklen_t,
-            )
-        } < 0
-        {
+        if unsafe { connect(sock, addr.as_ptr(), addr.len()) } < 0 {
             Err(Error::last_os_error())
         } else {
-            Ok(unsafe { VsockStream::from_raw_fd(sock) })
+            Ok(unsafe { Self::from_raw_fd(sock) })
         }
     }
 
     /// Open a connection to a remote host with specified cid and port.
     pub fn connect_with_cid_port(cid: u32, port: u32) -> Result<Self> {
-        Self::connect(&SockAddr::Vsock(VsockAddr::new(cid, port)))
+        Self::connect(&VsockAddr::new(cid, port))
     }
 
     /// Virtio socket address of the remote peer associated with this connection.
-    pub fn peer_addr(&self) -> Result<SockAddr> {
+    pub fn peer_addr(&self) -> Result<VsockAddr> {
         let mut vsock_addr = sockaddr_vm {
             svm_family: AF_VSOCK as sa_family_t,
             svm_reserved1: 0,
@@ -283,15 +263,12 @@ impl VsockStream {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr::new(
-                vsock_addr.svm_cid,
-                vsock_addr.svm_port,
-            )))
+            Ok(VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port))
         }
     }
 
     /// Virtio socket address of the local address associated with this connection.
-    pub fn local_addr(&self) -> Result<SockAddr> {
+    pub fn local_addr(&self) -> Result<VsockAddr> {
         let mut vsock_addr = sockaddr_vm {
             svm_family: AF_VSOCK as sa_family_t,
             svm_reserved1: 0,
@@ -310,10 +287,7 @@ impl VsockStream {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr::new(
-                vsock_addr.svm_cid,
-                vsock_addr.svm_port,
-            )))
+            Ok(VsockAddr::new(vsock_addr.svm_cid, vsock_addr.svm_port))
         }
     }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -58,7 +58,7 @@ impl VsockListener {
     /// Create a new VsockListener which is bound and listening on the socket address.
     pub fn bind(addr: &SockAddr) -> Result<VsockListener> {
         let mut vsock_addr = if let SockAddr::Vsock(addr) = addr {
-            addr.0
+            addr.as_ref().to_owned()
         } else {
             return Err(Error::new(
                 ErrorKind::Other,
@@ -116,7 +116,10 @@ impl VsockListener {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
+            Ok(SockAddr::Vsock(VsockAddr::new(
+                vsock_addr.svm_cid,
+                vsock_addr.svm_port,
+            )))
         }
     }
 
@@ -229,7 +232,7 @@ impl VsockStream {
     /// Open a connection to a remote host.
     pub fn connect(addr: &SockAddr) -> Result<Self> {
         let vsock_addr = if let SockAddr::Vsock(addr) = addr {
-            addr.0
+            addr.as_ref()
         } else {
             return Err(Error::new(
                 ErrorKind::Other,
@@ -244,7 +247,7 @@ impl VsockStream {
         if unsafe {
             connect(
                 sock,
-                &vsock_addr as *const _ as *const sockaddr,
+                vsock_addr as *const _ as *const sockaddr,
                 size_of::<sockaddr_vm>() as socklen_t,
             )
         } < 0
@@ -280,7 +283,10 @@ impl VsockStream {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
+            Ok(SockAddr::Vsock(VsockAddr::new(
+                vsock_addr.svm_cid,
+                vsock_addr.svm_port,
+            )))
         }
     }
 
@@ -304,7 +310,10 @@ impl VsockStream {
         {
             Err(Error::last_os_error())
         } else {
-            Ok(SockAddr::Vsock(VsockAddr(vsock_addr)))
+            Ok(SockAddr::Vsock(VsockAddr::new(
+                vsock_addr.svm_cid,
+                vsock_addr.svm_port,
+            )))
         }
     }
 

--- a/tests/vsock.rs
+++ b/tests/vsock.rs
@@ -17,7 +17,7 @@
 use rand::RngCore;
 use sha2::{Digest, Sha256};
 use std::io::{Read, Write};
-use vsock::{get_local_cid, SockAddr, VsockAddr, VsockStream, VMADDR_CID_HOST};
+use vsock::{get_local_cid, VsockAddr, VsockStream, VMADDR_CID_HOST};
 
 const TEST_BLOB_SIZE: usize = 1_000_000;
 const TEST_BLOCK_SIZE: usize = 5_000;
@@ -39,8 +39,7 @@ fn test_vsock() {
     rx_blob.resize(TEST_BLOB_SIZE, 0);
     rng.fill_bytes(&mut blob);
 
-    let mut stream =
-        VsockStream::connect(&SockAddr::Vsock(VsockAddr::new(3, 8000))).expect("connection failed");
+    let mut stream = VsockStream::connect(&VsockAddr::new(3, 8000)).expect("connection failed");
 
     while tx_pos < TEST_BLOB_SIZE {
         let written_bytes = stream

